meter/goodwe/server.go

```diff
@@ -67,43 +74,43 @@ func (m *Server) readData() {
 			return
 		}
 	}
+
 	m.readData()
 }
 
 func (m *Server) listen() {
 	for {
 		buf := make([]byte, 1024)
-		_, addr, err := m.conn.ReadFromUDP(buf)
+		n, addr, err := m.conn.ReadFromUDP(buf)
 		if err != nil {
+			m.log.ERROR.Println(err)
 			continue
 		}
+		m.log.TRACE.Printf("recv from %s: % X", addr, buf[:n])
 
 		ip := addr.IP.String()
-
-		mu.Lock()
-		if buf[4] == 250 {
-			vPv1 := float64(int16(binary.BigEndian.Uint16(buf[11:]))) * 0.1
-			vPv2 := float64(int16(binary.BigEndian.Uint16(buf[19:]))) * 0.1
-			iPv1 := float64(int16(binary.BigEndian.Uint16(buf[13:]))) * 0.1
-			iPv2 := float64(int16(binary.BigEndian.Uint16(buf[21:]))) * 0.1
-			iBatt := float64(int16(binary.BigEndian.Uint16(buf[167:]))) * 0.1
-			vBatt := float64(int16(binary.BigEndian.Uint16(buf[165:]))) * 0.1
-
-			pvPower := vPv1*iPv1 + vPv2*iPv2
-
-			inverter := m.inverters[ip]
-			inverter.PvPower = pvPower
-			inverter.BatteryPower = vBatt * iBatt
-			inverter.NetPower = -float64(int32(binary.BigEndian.Uint32(buf[83:])))
-
-			m.inverters[ip] = inverter
+		monitor := m.GetInverter(ip)
+		if monitor == nil {
+			m.log.ERROR.Println("unknown inverter:", ip)
+			continue
 		}
 
-		if buf[4] == 26 {
-			inverter := m.inverters[ip]
-			inverter.Soc = float64(buf[20])
-			m.inverters[ip] = inverter
-		}
-		mu.Unlock()
+		monitor.SetFunc(func(inverter Inverter) Inverter {
+			if buf[4] == 250 {
+				ui := func(u, i int16) float64 {
+					return float64(int16(binary.BigEndian.Uint16(buf[u:]))) *
+						float64(int16(binary.BigEndian.Uint16(buf[i:]))) / 100
+				}
+				inverter.PvPower = ui(11, 19) + ui(13, 21)
```

Comment:
Hier sind die indexe vertauscht. uPv1 = 11 & iPv1 = 13; uPv2 = 19 & iPv2 = 21.

```suggestion
		inverter.PvPower = ui(11, 13) + ui(19, 21)
```